amqp/src/main/java/io/github/tcdl/msb/adapters/amqp/AmqpMessageConsumer.java

```diff
@@ -41,27 +43,34 @@ public AmqpMessageConsumer(Channel channel, ExecutorService consumerThreadPool,
 
     @Override
     public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
+        long deliveryTag = envelope.getDeliveryTag();
+        AmqpAcknowledgementHandler ackHandler = createAcknowledgementHandler(
+                getChannel(), consumerTag, deliveryTag, amqpBrokerConfig.isRequeueRejectedMessages());
         try {
             Charset charset = amqpBrokerConfig.getCharset();
-            boolean requeueRejectedMessages = amqpBrokerConfig.isRequeueRejectedMessages();
+
             String bodyStr = new String(body, charset);
+
             LOG.debug(String.format("[consumer tag: %s] Message consumed from broker: %s", consumerTag, bodyStr));
+
             try {
-                consumerThreadPool.submit(new AmqpMessageProcessingTask(consumerTag, bodyStr, getChannel(), envelope.getDeliveryTag(), msgHandler));
-                LOG.debug(String.format("[consumer tag: %s] Message has been put in the processing queue: %s. About to send AMQP ack...",
+                consumerThreadPool.submit(new AmqpMessageProcessingTask(consumerTag, bodyStr, msgHandler, ackHandler));
+                LOG.debug(String.format("[consumer tag: %s] Message has been put in the processing queue: %s.",
                         consumerTag, bodyStr));
-
-                getChannel().basicAck(envelope.getDeliveryTag(), false);
```

Comment:
It seems like acknowledgement logic is changed here and you auto-ack messages not when they are consumed from queue but when the task from thread pool gets executed. I think it is really cool but you probably need to change the logic for the shutdown of the thread pool. You no longer need to way until all tasks from the queue complete. It's sufficient to wait for the completion of only the running tasks.